### PR TITLE
Adding a magic header to the file

### PR DESCRIPTION
**Pull request by Yoric**

### CODE REVIEW

File: format.py [hunk 1]
```diff
@@ -35,17 +45,23 @@ def write(types, string_dict, ty, tree, out):
...
+  content_no_brotli = io.BytesIO()
+
+  # Collect the local strings
   local_strings = strings.StringCollector(types)
   local_strings.visit(ty, tree)
   local_strings.strings -= set(string_dict)
   local_strings = list(sorted(local_strings.strings))
   string_dict = local_strings + string_dict
-  strings.write_dict(out, local_strings, with_signature=False)
+  strings.write_dict(content_no_brotli, local_strings, with_signature=False)
```
Comment (dominiccooney) [inline format.py:60]: Could you leave this as "out"? It doesn't seem relevant to the file header.

File: format.py [hunk 2]
```diff
@@ -126,15 +145,25 @@ def read(types, string_dict, ty, inp):
   >>> assert json.dumps(tree_in) == json.dumps(tree_out)
   '''
 
-  # Read the local string table
-  local_strings = strings.read_dict(inp, with_signature=False)
+  # Check that we're attempting to decompress a file with the right type.
```
Comment (dominiccooney) [inline format.py:148]: We don't want to do this where we embed Brotli into the file format. It's simpler to apply Brotli over the content of the whole file.

**Review by dominiccooney — CHANGES_REQUESTED** [review]

File: format.py (hunk 2 above)
Comment (Yoric) [reply format.py:148]: Do you see another way to add a magic header? The browser needs the ability to sniff that the file is binast, rather than brotli.

Comment (dominiccooney) [reply format.py:148]: Brotli is a content encoding, not a content type. The browser needs to sniff the content type. Putting a string preamble which isn't valid JavaScript should be sufficient.

Comment (Yoric) [reply format.py:148]: The content type is `application/javascript-binast`. Once we land our patches, there will be two content encodings for BinAST: context-0.1 (used by Facebook) and multipart (used by Cloudflare, possibly also by Facebook?). We need to detect between them quickly, and the simplest way to do it is a magic header.

There are alternatives (e.g. `Content-Encoding=context-0.1;brotli`), but it's not clear how to specify them (the list of content encodings is theoretically normalized) and we're on a deadline, so the consensus until now was to add a magic header.

I'm entirely ok with revising this choice in 0.2, once we have a little breathing time.

Comment (dominiccooney) [reply format.py:148]: I agree that a header makes it easy to detect which version of the format the browser is handling, and that we should add a header.

However the "consensus" you mention does not extend to me and embedding Brotli in the file format in this way. It would be a mistake or misrepresentation to suggest there was consensus about that aspect. Even in the context pure expediency for a serving experiment, embedding Brotli in the format in this way is unusual on our serving infrastructure and goes against the idea of making minimal changes to this format so we can get that experiment done.

I suggest that you let the browser's network stack handle the content encoding, and check the header post-decompression in the application/javascript-binast handler, maybe in frontend::CompileGlobalBinASTScript, to determine the version of the file format. This is simpler for authors because it leaves the BinAST format and content encoding orthogonal. I suspect this is also simple for browsers because content type and encoding are typically orthogonal in this way. If that's not the case for Firefox I would be curious to see pointers to the code so we can understand why.

By the way, Facebook is serving the BinAST files for the first serving experiment with content-encoding: br, so if you add the version detection which looks at raw bytes, you may break that serving experiment.

Comment (vdjeric) [reply format.py:148]: A while back we talked about using temporary content types (e.g. applicaton/javascript-binaryast-test) for the field trials. Why don't we just use application/javascript-binaryast-test1 and application/javascript-binaryast-test2 for the different encodings?

Note also that the "multipart" field trial will be short-lived and likely only open to a small whitelist of devs at FB & Mozilla, so we don't have to worry about supporting it longer term or disambiguating between it and "context" format, we can just switch over to the "context" format for the second trial and pretend the old one never existed.

Comment (Yoric) [reply format.py:148]: > However the "consensus" you mention does not extend to me and embedding Brotli in the file format in this way.

I will answer this point on a less public space.

> [...]
> I suggest that you let the browser's network stack handle the content encoding, and check the header post-decompression in the application/javascript-binast handler, maybe in frontend::CompileGlobalBinASTScript, to determine the version of the file format.

Technically, not `frontend::CompileGlobalBinASTScript`, but that's not the issue. We really want to avoid this as this kills a number of benefits of lazy parsing.

> [...]
>
> By the way, Facebook is serving the BinAST files for the first serving experiment with content-encoding: br, so if you add the version detection which looks at raw bytes, you may break that serving experiment.

Well, we'll have to see what works. At this stage, I can make no promise.

> A while back we talked about using temporary content types (e.g. applicaton/javascript-binaryast-test) for the field trials. Why don't we just use application/javascript-binaryast-test1 and application/javascript-binaryast-test2 for the different encodings?

It's a hack, but it should work as throwaway code. This would give us the existing `application/javascript-binast` for multipart and something like `application/javascript-binast-test-cx-01` for context 0.1.

See my weekly report for an early estimate of the dev·time cost of this change, though.

> Note also that the "multipart" field trial will be short-lived and likely only open to a small whitelist of devs at FB & Mozilla, so we don't have to worry about supporting it longer term or disambiguating between it and "context" format, we can just switch over to the "context" format for the second trial and pretend the old one never existed.

Fair enough, but this has no impact on the browser, as we need to support the other partners, too.

Comment (vdjeric) [reply format.py:148]: We talked about this in the BinAST weekly today. We're going to rely on Content-Type and Content-Encoding response header. @Yoric let's close this pull request?

Comment (Yoric) [reply format.py:148]: Let's do that.
